### PR TITLE
feat: split orchestrator/daemon images, default Opus 4.7, fix PEM parsing

### LINKED ISSUE #7: chore(test): DI refactor for checkout.ts and executor.ts to enable module-level unit tests
## Context

FR-007 and FR-008 of the "Project Housekeeping" feature (PR #6) explicitly required module-level unit tests for `src/core/checkout.ts` and `src/core/executor.ts` at ≥90% per-file line coverage. Both were formally **deferred** during implementation and are tracked here as a follow-up.

Originating work:
- PR: #6
- Spec rationale: [`specs/20260409-081113-project-housekeeping/spec.md` § Clarifications Session 2026-04-10](https://github.com/chrisleekr/github-app-playground/blob/main/specs/20260409-081113-project-housekeeping/spec.md)
- Plan waiver: [`plan.md` § Constitution Check — V. Test Coverage PASS with documented waiver](https://github.com/chrisleekr/github-app-playground/blob/main/specs/20260409-081113-project-housekeeping/plan.md)
- Tracking task: T059 (Phase 8.E.5)

## Why the deferral

Two Bun runtime limitations make module-level mocking of these files impractical without a source refactor:

1. **`checkout.ts`** calls the Bun built-in `$` template tag for git operations. `mock.module("bun", ...)` **does not** intercept built-in modules — Bun exposes `$` directly from the runtime, not from a module graph the mock system can interpose. As a result, any unit test for `checkout.ts` would spawn real `git` subprocesses, violating the constitution's "no real external calls in tests" principle (V) and making test runs non-hermetic.

2. **`executor.ts`** depends on `@anthropic-ai/claude-agent-sdk`. Bun's `mock.module(...)` persists the mock globally across *all* test files in the same process run — there is no per-file isolation. This means any module-level mock placed in a dedicated `executor.test.ts` would bleed into `router.test.ts` (and vice versa), corrupting either test. The two cannot coexist without breaking each other.

Both modules ARE transitively exercised by `test/webhook/router.test.ts` where the mocks are already in place, and global line coverage remains >99%. The gap is narrow: isolated failure modes that only `checkout.ts` / `executor.ts` unit tests would catch (e.g. an `executor.ts` refactor that breaks its own error-path branching without affecting the caller).

## Acceptance criteria

- [ ] `src/core/checkout.ts` refactored so external collaborators (`$` / `fs`) are injected via a parameter or a factory, rather than imported at module scope. The injection point should be covered by a type so misuse is caught at compile time.
- [ ] `src/core/executor.ts` refactored so the `@anthropic-ai/claude-agent-sdk` dependency is injected rather than imported at module scope.
- [ ] New file `test/core/checkout.test.ts` with per-file line coverage ≥90% and function coverage ≥90%. Tests MUST cover: successful PR checkout, successful issue checkout, credential-helper mode 0o700, cleanup on clone failure, cleanup idempotency, edge case (empty branch).
- [ ] New file `test/core/executor.test.ts` with per-file line coverage ≥90% and function coverage ≥90%. Tests MUST cover: successful execution with result message, timeout handling (Promise.race), agent throws → `success: false`, cost/turn metadata extraction, Anthropic vs Bedrock provider env.
- [ ] `bun run check` passes with the existing `bunfig.toml` per-file threshold (`{ lines = 0.9, functions = 0.9 }`).
- [ ] `router.test.ts` continues to pass after the refactor (the router mocks that currently cover these modules can be relaxed or replaced once dedicated tests exist).
- [ ] Spec annotations on FR-007 and FR-008 updated: DEFERRED markers removed, waiver rationale deleted, follow-up reference to this issue replaced with "Closed by #<this-issue-number>".
- [ ] `plan.md` Constitution Check V entry updated from "PASS with documented waiver" back to "PASS".

## Out of scope

- Refactoring the broader core pipeline (only `checkout.ts` and `executor.ts` need DI for this issue).
- Replacing `mock.module` with a different test isolation strategy (Bun worker-based test runs, etc.) — fix the two modules in place.
- Adding integration tests that spawn real `git` or real Claude Agent SDK calls.

## Notes

- If upstream Bun adds `mock.module` interception for built-ins or per-file mock isolation before this issue is picked up, revisit — the DI refactor may become unnecessary.
- The DI pattern for `setInterval` + `bind()` used in `router.ts` (Phase 7.B) is a good reference for how to keep the call site ergonomic while allowing tests to inject collaborators.

### PR DESCRIPTION
**Pull request by chrisleekr**
## Description

Three bundled changes on `feat/daemon-image-split`:

1. **Dual-image split.** The monolithic `Dockerfile` is replaced by `Dockerfile.orchestrator` (lean webhook server) and `Dockerfile.daemon` (rich CLI toolchain + scope-narrowed `sudo apt-get` for the `bun` user so Claude can install ad-hoc tooling at job time). `scripts/check-dockerfile-base-sync.ts` is a new CI guard that fails the build if the two Dockerfiles' `# --- SHARED-BASE-BEGIN/END` regions drift. CI matrix expands to 4 legs (`{amd64, arm64} × {orchestrator, daemon}`) with runners pinned to `ubuntu-24.04` / `ubuntu-24.04-arm`; Trivy scans each variant separately; per-variant tags publish as `:<ver>-orchestrator` / `:<ver>-daemon` with a bare `:<ver>` compat alias on the orchestrator so existing consumers keep working.
2. **Default `CLAUDE_MODEL` to `claude-opus-4-7`** when `CLAUDE_PROVIDER=anthropic` and unset. Implemented as a schema-level Zod `.transform()` after `.superRefine()`, so `Config.model` is inferred as `string` — the stale `config.model !== undefined` guards in `src/core/executor.ts` and `src/k8s/job-spawner.ts` are removed. Bedrock path still requires an explicit `CLAUDE_MODEL`.
3. **Fix issue #7 (`ERR_OSSL_BAD_END_LINE`).** `GITHUB_APP_PRIVATE_KEY` loaded from a K8s Secret populated by the `"---BEGIN---\n...\n---END---"` single-line form decodes to literal backslash+n bytes, not `0x0a`. Normalize at config load with `?.replace(/\\n/g, "\n")` — idempotent for real-newline PEMs.

Secondary: `src/daemon/tool-discovery.ts` splits into static (baked at image build via `scripts/generate-capabilities-manifest.ts` → `/app/daemon-capabilities.static.json`) vs dynamic (resources, auth, hostname, cached repos, container-daemon-running) probes; tool list extended to ~50 entries; parallel `which` probes; `VERSION_FLAGS` override map for tools whose `--version` is wrong/slow. `src/core/prompt-builder.ts` emits `Bash(sudo apt-get:*)` and an "On-Demand Package Installation" section when `daemonCapabilities` is present.

### Before / After (image layout)

```mermaid
flowchart LR
    MONO["chrisleekr/github-app-playground:&lt;tag&gt;<br/>single image, docker CLI, no rich toolchain"]:::before
    MONO --> OPS["Ops: webhook server"]:::before
    MONO --> WORK["Ops: daemon worker"]:::before

    ORCH["...:&lt;tag&gt;-orchestrator<br/>+ bare :&lt;tag&gt; compat alias<br/>lean, ~200-300 MB"]:::after
    DAEM["...:&lt;tag&gt;-daemon<br/>rich toolchain + sudo apt-get, ~2-3 GB"]:::after
    ORCH --> OPS2["Helm: orchestrator Deployment"]:::after
    DAEM --> WORK2["Helm: daemon Deployment"]:::after

    classDef before fill:#fee2e2,stroke:#991b1b,color:#111827
    classDef after  fill:#d1fae5,stroke:#065f46,color:#111827
```

## Related Issues

- Closes #7

## Testing

- [x] I have tested these changes locally
- [x] I have added/updated tests as needed
- [x] All existing tests pass

Local verification:

- `bun run typecheck` — green
- `bun run format` — green
- `bun run check:dockerfile-base-sync` — green (SHARED-BASE blocks byte-identical)
- `bun test` — 256 tests pass

Not yet verified in this PR (follow-up): actual `docker build` of both images (runs in CI) and homelab ArgoCD sync.

## Screenshots/Recordings

N/A — no UI changes.

## Changes

- **New:** `Dockerfile.orchestrator`, `Dockerfile.daemon`, `scripts/check-dockerfile-base-sync.ts`, `scripts/generate-capabilities-manifest.ts`
- **Deleted:** `Dockerfile` (replaced by the two above)
- **`src/config.ts`:** schema-level `.transform()` defaults Anthropic `model` to `claude-opus-4-7`; `privateKey` normalizes `\n` escape sequences
- **`src/core/executor.ts`, `src/k8s/job-spawner.ts`:** drop stale `config.model !== undefined` guards
- **`src/daemon/tool-discovery.ts`:** static/dynamic capability split; baked manifest fast path; parallel probes; extended tool list; `VERSION_FLAGS` Map (not Record — sidesteps `eslint-plugin-security`)
- **`src/core/prompt-builder.ts`:** emit `Bash(sudo apt-get:*)` + add "On-Demand Package Installation" section for daemon mode
- **`.github/workflows/docker-build.yml`:** 4-leg `{arch, variant}` matrix, pinned runners, per-variant cache scope, per-variant digest artifact, per-variant merge job + Trivy scan, variant-suffix tag publishing
- **`.github/workflows/ci.yml`:** add `check:dockerfile-base-sync` step
- **`package.json`:** split `docker:build` into `docker:build:orchestrator` / `docker:build:daemon`; add `docker:run:daemon`; add `check:dockerfile-base-sync` script
- **Docs:** `.env.example`, `CLAUDE.md`, `docs/SETUP.md` updated for Opus 4.7 default + `ERR_OSSL_BAD_END_LINE` troubleshooting row

## Summary by CodeRabbit

* **New Features**
  * Two container variants (orchestrator, daemon) with local run scripts.
  * Daemon can perform limited on-demand package installs under scoped apt permissions.
  * Added a capabilities manifest generator and faster static+dynamic tool discovery.

* **Documentation**
  * Updated Anthropic default model to claude-opus-4-7 in docs and examples.
  * Added SSL private-key troubleshooting note.

* **Chores**
  * CI now includes a Dockerfile base-sync guard and multi-variant build/publish workflow.

### CODE REVIEW

File: .github/workflows/docker-build.yml [hunk 1]
```diff
@@ -202,26 +229,27 @@ jobs:
...
         run: docker buildx imagetools inspect "${IMAGE}:${TAG}"
 
+      - name: Emit per-variant tag for downstream jobs
+        id: emit
+        env:
+          VARIANT: ${{ matrix.variant }}
+          VARIANT_TAG: ${{ steps.tag.outputs.variant_tag }}
+        run: |
+          echo "${VARIANT}-tag=${VARIANT_TAG}" >> "$GITHUB_OUTPUT"
+
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply .github/workflows/docker-build.yml]: Fixed in 2422af2 — dissolved the job-level `outputs` on the merge job (which were non-deterministic when two matrix legs wrote the same key) and the separate "Emit per-variant tag" step. The downstream scan job now recomputes `${tag-name}-${matrix.variant}` locally, which is the same deterministic value each leg would have produced anyway.

File: src/core/prompt-builder.ts [hunk 2]
```diff
@@ -319,6 +319,11 @@ export function resolveAllowedTools(
       tools.push(`Bash(${daemonCapabilities.containerRuntime.name}:*)`);
     }
 
+    // Scope-narrowed sudoers rule in Dockerfile.daemon lets the `bun` user run
+    // ONLY `apt-get update` and `apt-get install -y <pkg>`; the allow-list
+    // matches that scope so the audit story is clean.
+    tools.push("Bash(sudo apt-get:*)");
+
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/core/prompt-builder.ts]: Fixed in cb49986 — replaced `Bash(sudo apt-get:*)` with three narrow patterns (`sudo apt-get update:*`, `sudo apt-get install -y:*`, `sudo apt-get install --no-install-recommends -y:*`) that mirror the sudoers `Cmnd_Alias` in Dockerfile.daemon.

File: src/core/prompt-builder.ts [hunk 3]
```diff
@@ -360,6 +365,19 @@ You are running on a daemon worker process with the following environment:
...
+This daemon has full autonomous \`apt-get\` access. If a tool you need is not in
+the baked inventory above, you MAY install it before proceeding:
+
+  sudo apt-get update
+  sudo apt-get install -y <package>
+
+Configured apt sources: Debian trixie, NodeSource, GitHub CLI, Microsoft
+(azure-cli), Charmbracelet, MongoDB, Google Cloud SDK. Install only packages
+required to complete the task; do not install speculative extras. Record any
+install in your tracking comment so reviewers see what was added at job-time.
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/core/prompt-builder.ts:390]: Fixed in cb49986 — the prompt text was reworded to describe the actual scoped access: only `apt-get update` and `apt-get install` against upstream package names from configured sources, matching what the sudoers rule now permits.

File: .github/workflows/docker-build.yml [hunk 4]
```diff
@@ -139,32 +162,39 @@ jobs:
...
+    runs-on: ubuntu-24.04
     timeout-minutes: 10
+    strategy:
+      fail-fast: false
+      matrix:
+        variant: [orchestrator, daemon]
     outputs:
-      tag: ${{ steps.tag.outputs.tag }}
+      orchestrator-tag: ${{ steps.emit.outputs.orchestrator-tag }}
+      daemon-tag: ${{ steps.emit.outputs.daemon-tag }}
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply .github/workflows/docker-build.yml]: Fixed in 2422af2 — same root cause as #3105022242. The scan job no longer depends on hyphenated matrix outputs; it computes `${tag-name}-${matrix.variant}` directly, so the hyphen-in-outputs-key GHA quirk no longer matters.

File: package.json [hunk 5]
```diff
@@ -59,8 +59,12 @@
...
-    "docker:build": "docker build -t chrisleekr/github-app-playground:local . --progress=plain",
-    "docker:run": "bun run docker:build && docker run -p 3000:3000 --env-file .env -v $HOME/.aws:/home/bun/.aws:ro --rm --name github-app-playground chrisleekr/github-app-playground:local",
+    "docker:build:orchestrator": "docker build -f Dockerfile.orchestrator -t chrisleekr/github-app-playground:local-orchestrator . --progress=plain",
+    "docker:build:daemon": "docker build -f Dockerfile.daemon -t chrisleekr/github-app-playground:local-daemon . --progress=plain",
+    "docker:build": "bun run docker:build:orchestrator && bun run docker:build:daemon",
+    "docker:run": "bun run docker:build:orchestrator && docker run -p 3000:3000 -p 3002:3002 --env-file .env -v $HOME/.aws:/home/bun/.aws:ro --rm --name github-app-playground chrisleekr/github-app-playground:local-orchestrator",
+    "docker:run:daemon": "bun run docker:build:daemon && docker run --env-file .env -e ORCHESTRATOR_URL=ws://host.docker.internal:3002 -v $HOME/.aws:/home/bun/.aws:ro --rm --name github-app-playground-daemon chrisleekr/github-app-playground:local-daemon",
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply package.json:66]: Fixed in 5f77cc6 — `JOB_IMAGE` default in src/k8s/job-spawner.ts updated from `github-app-playground:local` to `github-app-playground:local-orchestrator` so it matches the variant-suffixed tags the docker-build workflow now produces.

File: scripts/check-dockerfile-base-sync.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  const endIdx = contents.indexOf(END_MARKER);
+  if (beginIdx === -1) {
+    throw new Error(`${filePath}: missing '${BEGIN_MARKER}' marker`);
+  }
+  if (endIdx === -1) {
+    throw new Error(`${filePath}: missing '${END_MARKER}' marker`);
+  }
+  if (endIdx <= beginIdx) {
+    throw new Error(`${filePath}: '${END_MARKER}' must appear after '${BEGIN_MARKER}'`);
+  }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply scripts/check-dockerfile-base-sync.ts:40]: Fixed in 3edd206 — added duplicate-marker guards after the existing begin/end index lookups. If either marker appears more than once in a Dockerfile, the check now throws instead of silently comparing only the first pair.

File: src/daemon/tool-discovery.ts [hunk 7]
```diff
@@ -55,117 +179,202 @@ function discoverTool(name: string, versionFlag = "--version"): DiscoveredTool {
...
-// Ephemeral detection
-
-function detectEphemeral(): boolean {
-  // Docker / container detection
-  if (existsSync("/.dockerenv")) return true;
-  // Kubernetes pod
-  if (process.env["KUBERNETES_SERVICE_HOST"] !== undefined) return true;
-  return false;
+async function discoverToolsParallel(names: readonly string[]): Promise<DiscoveredTool[]> {
+  return Promise.all(names.map((n) => Promise.resolve(discoverTool(n))));
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/daemon/tool-discovery.ts]: Fixed in 7a2b50b — swapped `execFileSync` for `promisify(execFile)` throughout tool-discovery.ts. `discoverToolsParallel` now genuinely fans out via `Promise.all`, and the container-runtime probe runs in parallel with tool discovery in `probeStaticCapabilities`.

File: src/daemon/tool-discovery.ts [hunk 8]
```diff
@@ -30,20 +139,35 @@ export function getCurrentResources(): DaemonResources {
...
+  // nvm is a shell function sourced from /usr/local/nvm/nvm.sh, not a binary.
+  // `which nvm` always fails in a non-interactive shell, so fall back to a
+  // filesystem check when the script is installed at the baked path.
+  if (name === "nvm") {
+    const nvmScript = "/usr/local/nvm/nvm.sh";
+    if (existsSync(nvmScript)) {
+      return { name, path: nvmScript, version: "sourced", functional: true };
+    }
+    return { name, path: "", version: "", functional: false };
+  }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/daemon/tool-discovery.ts:166]: Fixed in 7a2b50b — same migration. `discoverTool`, `probeContainerRuntimeStatic`, `probeContainerDaemonRunning`, and `probeAuthContexts` are all async now, so no subprocess blocks the event loop.

File: src/daemon/tool-discovery.ts [hunk 9]
```diff
@@ -55,117 +179,202 @@ function discoverTool(name: string, versionFlag = "--version"): DiscoveredTool {
...
+      ephemeral,
+      staticManifest: bakedManifest !== null,
+    },
+    "Capability discovery complete",
+  );
+
+  // Validate the merged shape — catches drift between the baked manifest and
+  // the schema after a daemon-types.ts change.
+  return daemonCapabilitiesSchema.parse(capabilities);
 }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/daemon/tool-discovery.ts]: Fixed in 7a2b50b — added `nvm` to `CLI_TOOL_NAMES` so the daemon probes for and reports it in the capability manifest.

File: .env.example [hunk 10]
```diff
@@ -24,10 +24,12 @@ ANTHROPIC_API_KEY=
 # CLAUDE_CODE_OAUTH_TOKEN=
 
 # Model override. Required when CLAUDE_PROVIDER=bedrock (Bedrock uses a different model ID
-# format than the Anthropic API). Optional for anthropic (SDK uses its own default).
+# format than the Anthropic API). Optional for anthropic — defaults to claude-opus-4-7
+# when unset. Opus runs ~5x Sonnet cost per token; set CLAUDE_MODEL=claude-sonnet-4-6
+# if cost-sensitive.
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply .env.example]: Fixed in 88870f6 — removed the `~5x Sonnet cost per token` claim from .env.example. Per Anthropic's published pricing, Opus 4.7 is ~1.67x Sonnet 4.6 per token, not ~5x. Rather than replace it with another number that'll drift with pricing changes, the comparison was dropped entirely.

File: CLAUDE.md [hunk 11]
```diff
@@ -72,6 +72,8 @@ The runtime bot in `src/` supports three authentication modes (see `src/config.t
 2. **`CLAUDE_CODE_OAUTH_TOKEN`** — Max/Pro subscription OAuth token (`sk-ant-oat...`, generated via `claude setup-token`). **Requires `ALLOWED_OWNERS`** to be set to a single-tenant value, because the [Agent SDK Note](https://code.claude.com/docs/en/agent-sdk/overview) prohibits serving other users' repos from a personal subscription quota. The token is forwarded to the Claude CLI subprocess via `buildProviderEnv()` in `src/core/executor.ts`; the CLI's own [auth precedence chain](https://code.claude.com/docs/en/authentication#authentication-precedence) picks between credentials if multiple are set.
 3. **AWS Bedrock** — full credential chain via `CLAUDE_PROVIDER=bedrock` + `AWS_REGION` + `CLAUDE_MODEL` (Bedrock model ID format). Credential resolution handled by the AWS SDK inside the subprocess.
 
+Default agent execution model when `CLAUDE_MODEL` is unset and `CLAUDE_PROVIDER=anthropic`: `claude-opus-4-7` (Opus 4.7). Opus runs ~5× Sonnet cost per token — set `CLAUDE_MODEL=claude-sonnet-4-6` if cost-sensitive. The Bedrock path still requires an explicit `CLAUDE_MODEL` (Bedrock model IDs differ from Anthropic's).
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply CLAUDE.md]: Fixed in 88870f6 — same fix as #3105951497. The `~5x Sonnet cost per token` sentence was removed from CLAUDE.md (and mirrored in src/config.ts) because the real ratio is closer to 1.67x and the exact multiplier adds no decision value.

File: Dockerfile.daemon [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+  fi
+
+# 11. Sudoers carve-out: full-autonomous apt access for the `bun` user.
+#     Scope-narrowed to `apt-get update` + `apt-get install -y <pkg>` — no
+#     remove, no purge, no arbitrary sudo. `visudo -cf` validates the file
+#     and fails the build if the rule is malformed.
+RUN echo 'bun ALL=(ALL) NOPASSWD: /usr/bin/apt-get update, /usr/bin/apt-get install -y *, /usr/bin/apt-get install --no-install-recommends -y *' \
+  > /etc/sudoers.d/bun-apt \
+  && chmod 0440 /etc/sudoers.d/bun-apt \
+  && visudo -cf /etc/sudoers.d/bun-apt
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply Dockerfile.daemon]: Fixed in cb49986 — the wildcard sudoers rule is replaced with a `Cmnd_Alias` that matches only Debian-Policy package names (`[a-z0-9][a-z0-9+.-]*`). Character-class excludes `/` in the name body and requires an alphanumeric first char, so local .deb paths (`./pkg.deb`, `/tmp/pkg.deb`) and URLs are rejected. Only `apt-get update` and `apt-get install -y/--no-install-recommends -y <name>` against configured upstream sources are permitted.

File: Dockerfile.orchestrator [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  apt-get update && \
+  apt-get install -y --no-install-recommends nodejs && \
+  rm -rf /var/lib/apt/lists/*
+
+# Base-image CVE patches: oven/bun:1.3.12 bakes in openssl 3.5.4-1~deb13u2.
+# Targeted upgrade keeps rebuilds reproducible; CI Trivy gate catches new flags.
+RUN apt-get update && \
+  apt-get upgrade -y --no-install-recommends \
+  openssl libssl3t64 openssl-provider-legacy && \
+  rm -rf /var/lib/apt/lists/*
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply Dockerfile.orchestrator:36]: Fixed in cb49986 — switched `apt-get upgrade -y openssl libssl3t64 openssl-provider-legacy` to `apt-get install --only-upgrade -y <same packages>`. `apt-get upgrade` ignores positional arguments and upgrades everything installed; `install --only-upgrade` actually respects the listed package names. Mirrored into Dockerfile.daemon's SHARED-BASE region so the sync check still passes.

File: src/core/prompt-builder.ts [hunk 14]
```diff
@@ -319,6 +319,11 @@ export function resolveAllowedTools(
       tools.push(`Bash(${daemonCapabilities.containerRuntime.name}:*)`);
     }
 
+    // Scope-narrowed sudoers rule in Dockerfile.daemon lets the `bun` user run
+    // ONLY `apt-get update` and `apt-get install -y <pkg>`; the allow-list
+    // matches that scope so the audit story is clean.
+    tools.push("Bash(sudo apt-get:*)");
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/core/prompt-builder.ts]: Fixed in cb49986 — the allow-list in prompt-builder.ts now uses three narrow patterns (`sudo apt-get update:*`, `sudo apt-get install -y:*`, `sudo apt-get install --no-install-recommends -y:*`) that mirror exactly what the sudoers `Cmnd_Alias` permits. No more `Bash(sudo apt-get:*)` catch-all.

File: .github/workflows/docker-build.yml [hunk 15]
```diff
@@ -42,45 +53,58 @@ on:
...
+              variant: daemon,
+              dockerfile: Dockerfile.daemon,
+            }
+          - {
+              platform: linux/arm64,
+              runner: ubuntu-24.04-arm,
+              arch: arm64,
+              variant: daemon,
+              dockerfile: Dockerfile.daemon,
+            }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply .github/workflows/docker-build.yml:97]: Validated — the workflow header (lines 13–17) documents the `ubuntu-24.04` pin as deliberate ("Runners are explicitly pinned (not `ubuntu-latest`) so the rolling alias can't silently flip to 26.04 and break the build"), so the guideline was stale, not the code. Synced `CLAUDE.md` lines 99 + 105 to match in 5989b41.

File: Dockerfile.daemon [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+# hadolint ignore=DL4001
+RUN set -eux; \
+  if echo "${INSTALL_LANGS}" | grep -qw go; then \
+  curl -fsSL "https://go.dev/dl/go${GO_VERSION}.linux-${TARGETARCH}.tar.gz" \
+  | tar -xz -C /usr/local; \
+  fi; \
+  if echo "${INSTALL_LANGS}" | grep -qw rust; then \
+  curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs \
+  | sh -s -- -y --default-toolchain stable --profile minimal; \
+  fi
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply Dockerfile.daemon:205]: Valid — install steps run as root, final `USER bun` (line 283) can't write `/usr/local/{cargo,rustup,pipx}` so `cargo install`, `rustup update`, `pipx install`, cargo registry fetches, and `nvm install` all EACCES. Fixed in f67e881 with a `chown -R bun:bun` over `/usr/local/pipx`, `/usr/local/cargo`, `/usr/local/rustup`, and `/usr/local/nvm`, guarded with `[ -d "$d" ]` so gated dirs (rust off in `INSTALL_LANGS`) don't fail the build.

File: Dockerfile.daemon [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+LABEL maintainer="Chris Lee"
+LABEL com.chrisleekr.bot.package-version=${PACKAGE_VERSION}
+LABEL com.chrisleekr.bot.git-hash=${GIT_HASH}
+LABEL com.chrisleekr.bot.variant=daemon
+
+# Docker CLI — Claude agents shell `docker build`/`docker compose` against a
+# dind sidecar. Matches src/k8s/job-spawner.ts major pin for client/daemon
+# compatibility. docker:29-cli ships Go 1.26.x which patches CVE-2025-47907 +
+# CVE-2026-32282 (net/http, archive/tar, database/sql).
+COPY --from=docker:29-cli /usr/local/bin/docker /usr/local/bin/docker
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply Dockerfile.daemon]: Valid — confirmed the bake order: `probeContainerRuntimeStatic()` in `src/daemon/tool-discovery.ts:197-215` calls `discoverTool("docker")` which needs the binary on PATH, and `discoverCapabilities()` (lines 319–346) merges the baked `staticCaps.containerRuntime` verbatim with no fallback re-probe — so the manifest would have baked `containerRuntime: null` and stuck with it at runtime despite the CLI being in the final image. Moved the `COPY --from=docker:29-cli` step into the `daemon-tools` stage before the manifest-bake `RUN` in c4b7e03.

File: Dockerfile.daemon [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+  curl -fsSL "https://dl.filippo.io/mkcert/${MKCERT_VERSION}?for=linux/${TARGETARCH}" \
+  -o /usr/local/bin/mkcert && chmod +x /usr/local/bin/mkcert
+
+# 5. AWS CLI v2 — official installer, arch-aware.
+# hadolint ignore=DL4001
+RUN set -eux; \
+  AWS_ARCH="x86_64"; [ "${TARGETARCH}" = "arm64" ] && AWS_ARCH="aarch64" || true; \
+  curl -fsSL "https://awscli.amazonaws.com/awscli-exe-linux-${AWS_ARCH}-${AWSCLI_VERSION}.zip" \
+  -o /tmp/awscliv2.zip && unzip -q /tmp/awscliv2.zip -d /tmp && \
+  /tmp/aws/install && rm -rf /tmp/awscliv2.zip /tmp/aws
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply Dockerfile.daemon:180]: Won't fix — accepting risk. Adding SHA256 verification for 14 tools across two arches has high maintenance overhead (per-tool checksum formats differ: `.sha256` / `.sha256sum` / `SHA256SUMS` / `checksums.txt`; shellcheck and mkcert publish no sidecar). Version pins + HTTPS from vendor-canonical origins is the accepted mitigation. Revisit in a dedicated hardening pass if threat model shifts.

File: Dockerfile.daemon [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+
+# 13. Bake the static capability manifest for fast daemon startup.
+#     Must run AFTER every tool install above. Consumed at daemon startup by
+#     src/daemon/tool-discovery.ts → loadStaticManifest().
+COPY package.json bun.lock tsconfig.json ./
+COPY src/ ./src/
+COPY scripts/generate-capabilities-manifest.ts ./scripts/generate-capabilities-manifest.ts
+RUN bun install --frozen-lockfile --production --ignore-scripts \
+  && bun run scripts/generate-capabilities-manifest.ts /app/daemon-capabilities.static.json \
+  && rm -rf node_modules scripts/generate-capabilities-manifest.ts src/ package.json bun.lock tsconfig.json
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply Dockerfile.daemon:270]: Fixed in de45b73. `COPY --from=docker:29-cli /usr/local/libexec/docker/cli-plugins/` added alongside the CLI binary copy so both `docker-compose` and `docker-buildx` plugins land in the image before `probeContainerRuntimeStatic()` runs during manifest bake. `composeAvailable: true` will now be baked correctly.

File: src/daemon/tool-discovery.ts [hunk 20]
```diff
@@ -1,19 +1,139 @@
...
+];
+
+/**
+ * Override for tools whose `--version` doesn't exist, takes too long, or emits
+ * noise. Anything not listed uses the default `--version`. Map (not Record) to
+ * sidestep eslint-plugin-security's object-injection rule on dynamic-key reads.
+ */
+const VERSION_FLAGS = new Map<string, string[]>([
+  ["kubectl", ["version", "--client=true", "--output=yaml"]],
+  ["helm", ["version", "--short"]],
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/daemon/tool-discovery.ts:130]: Fixed in de45b73. Dropped `--output=yaml` from the kubectl `VERSION_FLAGS` entry so plain output flows into the parser. Also tightened the parser itself: replaced `replace(/^[^0-9]*/, "")` with a bounded semver capture `v?(\d+\.\d+(?:\.\d+)?(?:[-.+][\w.-]{0,32})?)`, which handles the full list of probed tools (kubectl, helm, node, terraform, argocd, etc.) including pre-release/build suffixes like `v2.13.1+abc`.
